core/src/apps/cardano/address.py

```diff
@@ -1,54 +1,104 @@
 from trezor import log
 from trezor.crypto import base58, crc, hashlib
 
+from apps.cardano.protocol_magics import ProtocolMagics
 from apps.common import HARDENED, cbor
 from apps.common.seed import remove_ed25519_prefix
 
+if False:
+    from typing import Tuple
+    from trezor.crypto import bip32
+    from apps.cardano import seed
 
-def _encode_address_raw(address_data_encoded):
+
+def _encode_address_raw(address_data_encoded: bytes) -> str:
     return base58.encode(
         cbor.encode(
             [cbor.Tagged(24, address_data_encoded), crc.crc32(address_data_encoded)]
         )
     )
 
 
-def derive_address_and_node(keychain, path: list):
+def derive_address_and_node(
+    keychain: seed.Keychain, path: list, protocol_magic: int
+) -> Tuple[str, bip32.Node]:
     node = keychain.derive(path)
 
-    address_payload = None
-    address_attributes = {}
+    address_attributes = get_address_attributes(protocol_magic)
 
-    address_root = _get_address_root(node, address_payload)
+    address_root = _get_address_root(node, address_attributes)
     address_type = 0
     address_data = [address_root, address_attributes, address_type]
     address_data_encoded = cbor.encode(address_data)
 
     return (_encode_address_raw(address_data_encoded), node)
 
 
-def is_safe_output_address(address) -> bool:
+def get_address_attributes(protocol_magic: int) -> dict:
+    # protocol magic is included in Byron addresses only on testnets
+    if protocol_magic == ProtocolMagics.MAINNET:
+        address_attributes = {}
+    else:
+        address_attributes = {2: cbor.encode(protocol_magic)}
+
+    return address_attributes
+
+
+def is_safe_output_address(address: str) -> bool:
     """
     Determines whether it is safe to include the address as-is as
     a tx output, preventing unintended side effects (e.g. CBOR injection)
     """
+    address_data_encoded = _get_address_data(address)
+    if address_data_encoded is None:
+        return False
+
+    return _encode_address_raw(address_data_encoded) == address
+
+
+def _get_address_data(address: str) -> bytes:
```

Comment:
It's returning encoded data because the `is_safe_output_address` uses this data cbor encoded to try and recreate the address. The `matches_with_procol_magic` decodes this data further.

Reply:
`raise INVALID_ADDRESS` added in this [commit](https://github.com/trezor/trezor-firmware/pull/1096/commits/6ca9b50238a76e0aedecc6f8af3a001045ddb606).

Reply:
Hm, I just realised I should probably rename the function to `_get_encoded_address_data` then. I will do that soon.

Reply:
right, I got confused because the address data is encoded twice, and in both cases the result is a 2-tuple

but yes, please rename the function